pymediainfo/__init__.py

```diff
@@ -346,11 +346,45 @@ def parse(cls, filename, library_file=None, cover_data=False,
                 )
             for option_name, option_value in mediainfo_options.items():
                 lib.MediaInfo_Option(handle, option_name, option_value)
-        if lib.MediaInfo_Open(handle, filename) == 0:
-            lib.MediaInfo_Close(handle)
-            lib.MediaInfo_Delete(handle)
-            raise RuntimeError("An error occured while opening {}"
-                    " with libmediainfo".format(filename))
+        try:
+            filename.seek(0, 2)
+            file_size = filename.tell()
+            filename.seek(0)
+        except AttributeError:  # filename is not a file-like object
+            file_size = None
+
+        if file_size is not None:  # We have a file-like object, use the buffer protocol:
+            lib.MediaInfo_Open_Buffer_Init(handle, file_size, 0)
+            while True:
+                buffer = filename.read(7 * 188)
+                if buffer:
+                    try:
+                        if ctypes.c_size_t(lib.MediaInfo_Open_Buffer_Continue(handle, buffer, len(buffer))).value & 0x08:
+                            break
+                    except ctypes.ArgumentError:
+                        raise ValueError("file should be opened in binary mode")
```

Comment:
I'm glad you asked, I spent quite some time choosing it ^^

The `TypeError` vs `ValueError`, here is ambiguous:
- We can say it's the right type (a file like object) but the wrong value (opened in text mode).
- We can say it's the wrong type (a `TextIOWrapper` instead of a `BufferedReader`).

For a normal user, the first one is easy to understand, the doc only speak about "file like objects". Only by looking at the (current) implementation of `open` we spot those are different types.

In your example with `/dev/null` it's less ambigous : the `write` method on a `BufferedReader` expects a `bytes` object, if you give an `str` instead, it's the wrong type. (it's the argument to `write` that causes the exception, not the type of the file, even if by changing the type of the file the exception dissapear... I agree).

When writing it in the first place I remember digging and doodling to choose ValueError vs TypeError, and made my final decision by running:

```
$ git -C ../cpython grep 'should be opened in binary'
Lib/asyncio/base_events.py:            raise ValueError("file should be opened in binary mode")
[...]
Lib/socket.py:            raise ValueError("file should be opened in binary mode")
[...]
```

So choosing `ValueError` is consistent with the Python stdlib, and it's "developer-friendly" by saying "It's the right type: a file-like object, but the wrong value".

While doodling around it, I found another error, I only tested with a file readable in text mode, by adding a test with a binary file I get a UnicodeDecodeError on the read before.

Reply:
Thanks for the detailed explanation, let's keep it this way then :smiley:!